comp/logs/agent/agent.go

```diff
@@ -0,0 +1,84 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2023-present Datadog, Inc.
+
+package agent
+
+import (
+	"context"
+
+	configComponent "github.com/DataDog/datadog-agent/comp/core/config"
+	logComponent "github.com/DataDog/datadog-agent/comp/core/log"
+	"github.com/DataDog/datadog-agent/pkg/logs"
+	"github.com/DataDog/datadog-agent/pkg/logs/diagnostic"
+	"github.com/DataDog/datadog-agent/pkg/logs/schedulers"
+	"go.uber.org/fx"
+)
+
+type dependencies struct {
+	fx.In
+
+	Lc     fx.Lifecycle
+	Log    logComponent.Component
+	Config configComponent.Component
+}
+
+type agent struct {
+	logsAgent *logs.Agent
+	log       logComponent.Component
+	config    configComponent.Component
+}
+
+func newLogsAgent(deps dependencies) Component {
+	logsAgent := &agent{log: deps.Log, config: deps.Config}
+	deps.Lc.Append(fx.Hook{
+		OnStart: logsAgent.start,
+		OnStop:  logsAgent.stop,
+	})
+	return logsAgent
+}
+
+func (a *agent) start(context.Context) error {
+
+	if a.config.GetBool("logs_enabled") || a.config.GetBool("log_enabled") {
```

Comment:
What is our recommended pattern for components that satisfy the following criteria: 

- It can be started/stopped 
- it can fail to start because of an error
- it can fail to start because it's disabled at a config level. 

I am wondering because in order to use the start hook, the component instance has to "exist" and therefore cannot be optional. I suppose we could retroactively make the optional nil if `start` fails, but that could cause other problems with any downstream components.  

So by making this component optional we would have to remove the start hook and try to start the logs agent when it's constructed. 

Does that make sense? Or am I misunderstanding.

Reply:
If the log agent is disabled then the `start` method doesn't have to be called, right?
If the `start` method failed, then fx will stop the application.
What about something like:
```go
 type Component Optional[MyComponent]

type MyComponent interface {
	// Methods provided by the component
}

type logsAgent  struct {
	// Some fields
}

func newLogsAgent(deps dependencies) Component {
	if !config.GetBool("logs_enabled") {
		// By default the component is empty.
		// Optional.GetValue() returns (MyComponent, false)
		return Component{}
	}
        logsAgent := &logsAgent{log: deps.Log, config: deps.Config}     
        deps.Lc.Append(fx.Hook{
		OnStart: logsAgent.start,
		OnStop:  logsAgent.stop,
	}) 
	return Component(NewOptional[MyComponent](component))
}
```
Did I misunderstand something?

Reply:
I updated the PR to use optional however `type Component Optional[MyComponent]` will not work cleanly without having to convert it back to an optional with a type assertion, so instead I used an explicit optional everywhere.